vhost-device-spi/src/spi.rs

```diff
@@ -0,0 +1,1119 @@
+        }
+    }
+
+    fn get_bits_per_word(&self) -> Result<u8> {
+        let mut bpw: u8 = 0;
+
+        // SAFETY: Safe as the file is a valid SPI controller.
+        let ret = unsafe {
+            ioctl_with_mut_ref(&self.file.as_raw_fd(), SPI_IOC_RD_BITS_PER_WORD(), &mut bpw)
+        };
+
+        if ret == -1 {
+            Err(Error::IoctlFailure("get_bits_per_word", IoError::last()))
+        } else {
+            Ok(bpw)
+        }
+    }
+
+    fn set_bits_per_word(&self, bpw: u8) -> Result<()> {
+        // SAFETY: Safe as the file is a valid SPI controller.
+        let ret =
+            unsafe { ioctl_with_ref(&self.file.as_raw_fd(), SPI_IOC_WR_BITS_PER_WORD(), &bpw) };
+
+        if ret == -1 {
+            Err(Error::IoctlFailure("set_bits_per_word", IoError::last()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn get_mode(&self) -> Result<u32> {
+        let mut mode: u32 = 0;
+
+        // SAFETY: Safe as the file is a valid SPI controller.
+        let ret =
+            unsafe { ioctl_with_mut_ref(&self.file.as_raw_fd(), SPI_IOC_RD_MODE32(), &mut mode) };
+
+        if ret == -1 {
+            Err(Error::IoctlFailure("get_mode", IoError::last()))
+        } else {
+            Ok(mode)
+        }
+    }
+
+    fn set_mode(&self, mode: u32) -> Result<()> {
+        // SAFETY: Safe as the file is a valid SPI controller.
+        let ret = unsafe { ioctl_with_ref(&self.file.as_raw_fd(), SPI_IOC_WR_MODE32(), &mode) };
+
+        if ret == -1 {
+            Err(Error::IoctlFailure("set_mode", IoError::last()))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn rdwr(&self, reqs: &mut [SpiTransReq]) -> Result<()> {
+        let mut msgs: Vec<SpiIocTransfer> = Vec::with_capacity(reqs.len());
+        let len = reqs.len();
+        let mut tx_buf_ptr: *mut u8;
+        let mut rx_buf_ptr: *mut u8;
+
+        let saved_mode: u32 = self.get_mode()?;
+        let mut trans_mode: u32 = saved_mode;
+
+        for req in reqs {
+            if req.tx_buf.is_empty() {
+                tx_buf_ptr = ptr::null_mut();
+            } else {
+                tx_buf_ptr = req.tx_buf.as_mut_ptr();
+            }
+
+            if req.rx_buf.is_empty() {
+                rx_buf_ptr = ptr::null_mut();
+            } else {
+                rx_buf_ptr = req.rx_buf.as_mut_ptr();
+            }
+
+            msgs.push(SpiIocTransfer {
+                tx_buf: tx_buf_ptr as u64,
+                rx_buf: rx_buf_ptr as u64,
+                len: req.trans_len,
+                speed_hz: req.speed_hz,
+                delay_usecs: req.delay_usecs,
+                bits_per_word: req.bits_per_word,
+                cs_change: req.cs_change,
+                tx_nbits: req.tx_nbits,
+                rx_nbits: req.rx_nbits,
+                word_delay_usecs: req.word_delay_usecs,
+                pad: 0,
+            });
+
+            if (req.mode & SPI_CPHA) == SPI_CPHA {
+                trans_mode |= LNX_SPI_CPHA;
```

Comment:
See if you can use bitflags for this stuff.

Reply:
I move these definitions to a separate file. But when I try to use bitflags, I got the error that they cannot execute bitwise operations with u32. So I reserve them as const definition

Reply:
You can do bitwise operations like this:

https://github.com/rust-vmm/vm-memory/blob/main/src/mmap_xen.rs#L446

Reply:
Thanks, it works.